### DISCUSSION TITLE
paddle训练识别模型微调时，如果某些易错字符增多10%样本，会不会影响其他字符的准确率？怎样避免其他字符受影响？

### DISCUSSION THREAD

**Discussion by nissansz** [PaddlePaddle/PaddleOCR#14303] [Q&A]
paddle训练识别模型微调时，如果某些易错字符增多10%样本，会不会影响其他字符的准确率？怎样避免其他字符受影响？

**Comment by GreatV** [comment]
在使用Paddle进行模型微调时，如果增加了某些易错字符的样本量（例如增加10%的样本），可能会对其他字符的识别准确率产生一定影响，主要原因如下：

### 影响分析
1. **类别不平衡问题**：
   - 增加某些字符的样本量后，这些字符的样本可能会占据整体训练数据集的更大比例。如果训练数据的类别分布不均衡，模型可能会倾向于更频繁出现的字符，从而降低对低频字符的关注，导致其他字符的识别准确率下降。

2. **模型过拟合特定字符**：
   - 增加某些字符的样本量可能会导致模型在这些字符上过拟合，而无法有效学习其他字符的特征。

3. **梯度冲突**：
   - 在多类别分类任务中，不同类别的样本可能会对模型参数的更新方向产生冲突。如果某些类别的样本过多，这些类别的梯度可能会主导优化过程，从而影响其他类别的学习。

---

### 避免影响其他字符的策略

1. **数据均衡处理**：
   - **动态采样**：在训练数据加载时，动态调整各类别的采样比例，使各字符的样本数量更加均匀。例如，如果某些字符样本较多，可以降低其采样概率，确保其他字符有足够的训练机会。
   - **数据扩增**：对低频字符进行数据扩增（如旋转、缩放、加噪声等），以增强这些字符的样本多样性和数量。

2. **损失函数加权**：
   - 为了避免模型过度关注某些字符，可以对损失函数进行加权。通过给低频字符的损失赋予更高的权重，从而平衡模型对不同字符的学习。

3. **冻结部分参数**：
   - 如果您只想优化某些易错字符的识别效果，可以尝试冻结模型中已经学会其他字符特征的层，只微调最后几层或特定的权重。

4. **知识蒸馏**：
   - 使用之前训练好的模型作为教师模型，通过知识蒸馏的方式，在微调时保证对其他字符的性能不会下降。具体方法是同时优化模型的预测输出与教师模型输出之间的KL散度。

5. **分阶段训练**：
   - 先用均衡样本（不增加易错字符样本）进行整体训练，然后在最后阶段引入增加后的易错字符样本进行微调。这种方式可以减少对其他字符的影响。

6. **多任务学习**：
   - 如果可能，可以设计一个多任务学习框架，让模型同时学习字符分类和其他辅助任务（如语言模型），从而增强模型的泛化能力。

---

### 实践中的具体调整建议
1. **修改数据加载器**：
   - 在PaddleOCR中，您可以通过调整`ratio_list`参数来控制不同数据源的采样比例。例如：
     ```yaml
     Train:
       dataset:
         name: SimpleDataSet
         data_dir: ./train_data
         label_file_list:
           - ./train_data/label.txt
         ratio_list: [0.9, 0.1]
     ```
     上述配置可以控制主数据集与新增易错字符样本数据集的采样比例为9:1。

2. **使用预训练模型微调**：
   - 通过加载PaddleOCR提供的预训练模型，并在特定字符场景的数据上进行微调，能够快速收敛并减少对其他字符的影响。
     ```bash
     python3 tools/train.py -c configs/rec/rec_chinese_common_train.yml \
         -o Global.pretrained_model=./pretrained_model
     ```

3. **调整学习率**：
   - 在微调时，建议使用较小的学习率（如0.0001），这样可以避免对已有权重的过度调整，从而减少对其他字符的影响。

---

### 参考资料
- [PaddleOCR 微调文档](https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.3/doc/doc_ch/finetune.md)
- [Discussion on Fine-tuning Specific Characters](https://github.com/PaddlePaddle/PaddleOCR/discussions/14155)

通过以上方法，您可以在增加易错字符样本的同时，尽量减少对其他字符识别准确率的影响。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Comment by nissansz** [comment]
冻结部分参数：

如果您只想优化某些易错字符的识别效果，可以尝试冻结模型中已经学会其他字符特征的层，只微调最后几层或特定的权重。

这个具体怎么冻结？

比如有几万个字符，除了指定一些字符，比如几百个，其余能不能全部冻结？

**Comment by nissansz** [comment]
![image](https://github.com/user-attachments/assets/d354acc8-3540-47d6-aa7f-9e3e23ceda48)

类似有些模糊的这种图片，有没有什么代码可以处理成清晰图片？
或者还是说有没有什么代码可以模拟生产这种样本？寻要加样本训练？

**Comment by SWHL** [comment]
自己多试试吧
